packages/loader/container-loader/src/container.ts

```diff
@@ -379,81 +379,104 @@ export class Container
 	extends EventEmitterWithErrorHandling<IContainerEvents>
 	implements IContainer, IContainerExperimental
 {
+	/**
+	 * Not to be called directly on this type.
+	 */
+	public async initialize(): Promise<void> {
+		assert(false, "Method will be provided based on create/load API that is used");
```

Comment:
The assert text is hard to parse - it becomes clearer what it means only by reading the rest of the code.
I think we can avoid the need to have it, and the need to expose container in weird states.

How about the following:

- create() -> ContainerShell, can register on events only and has only one method:
   - ContainerShell.start() -> Container

This clearly requires proxying event registrations from shell to real container underneath.
But this kind of structure completely eliminates the risk of calling some public APIs on Container while it's not initialized. It also completely separates the workflows, allowing to remove APIs from Container (like initialize(), maybe others) that should not be there.

Reply:
Problem I faced with that is, do you also keep the shell around and continue to use it for events? It will actually be the same instance as the container returned by start.